### ISSUE TITLE
Precision issue when applying to HF transformer

### ISSUE THREAD

**Issue by fingertap** [zhuzilin/ring-flash-attention#67]
I checked the precision of forward pass using ring attn. I see large differences (e.g., -15.93 vs -11.75) between ring attention and flash attention:

```text
tensor([[-16.5000, -14.4375, -13.3125, -12.8750, -15.9375, -16.3750, -11.7500,
         -11.5000, -15.3125, -13.8125, -12.9375, -12.5000, -18.0000, -16.5000,
         -14.4375, -13.3125, -16.3750, -12.2500, -10.5000, -14.4375, -13.5625,
         -13.0000, -11.9375, -12.3750, -16.5000, -20.6250, -20.1250, -12.7500,
         -15.3750, -13.7500, -19.8750, -12.5625]], device='cuda:0',
       dtype=torch.bfloat16)
tensor([[-16.5000, -14.3125, -13.1875, -12.7500, -11.7500, -16.2500, -11.8125,
         -11.5000, -10.8125, -14.5625, -12.6250, -11.6250, -10.5000, -16.5000,
         -14.3125, -13.1875, -14.1250, -11.2500, -11.8750, -15.0625, -13.3750,
         -12.9375, -11.8125, -12.4375, -12.3125, -12.8125, -13.1250, -12.9375,
         -10.5625, -15.2500, -19.7500, -12.6250]], device='cuda:0',
       dtype=torch.bfloat16)
```

The code to reproduce:

```python
import torch
import torch.distributed as dist
import torch.nn.functional as F
from functools import cached_property


def reset_ring_attn_position_ids(start, end, packed_seq_lens):
    """
    Calculate position ids for packed_seq_ids[start:end].
    For example, if the packed_seq_lens is [3, 2, 4, 1], start=2, end=8,
    the position ids will be [2, 0, 1, 0, 1, 2].

    Args:
        start: the start position
        end: the end position
        packed_seq_lens: the sequence lengths of packed sequences
    """
    position_ids = torch.zeros((1, end - start), dtype=torch.long, device=torch.cuda.current_device())
    offset = 0
    for seqlen in packed_seq_lens:
        seq_start = max(offset, start)
        seq_end = min(offset + seqlen, end)
        if seq_start < seq_end:
            position_ids[0, seq_start - start : seq_end - start] = torch.arange(seq_start - offset, seq_end - offset)

        offset += seqlen
        if offset >= end:
            break
    return position_ids


def update_ring_attn_params(packed_seq_lens, total_seq_len, group):
    from ring_flash_attn import update_ring_flash_attn_params

    cu_seqlens = torch.cumsum(
        torch.tensor(packed_seq_lens, device=torch.cuda.current_device(), dtype=torch.int32),
        dim=-1,
        dtype=torch.int32,
    )
    cu_seqlens = F.pad(F.pad(cu_seqlens, (1, 0), value=0), (0, 1), value=total_seq_len)

    update_ring_flash_attn_params(cu_seqlens[0], group)


def convert_ring_attn_params(sequences, attention_mask, packed_seq_lens, ring_attn_group):
    # each rank within the ring group will process sequences[start:end]
    ring_attn_rank = dist.get_rank(group=ring_attn_group)
    ring_attn_size = dist.get_world_size(group=ring_attn_group)
    total_seq_len = sequences.numel()
    local_seq_len = total_seq_len // ring_attn_size
    start, end = ring_attn_rank * local_seq_len, (ring_attn_rank + 1) * local_seq_len
    sequences = sequences[:, start:end]
    attention_mask = attention_mask[:, start:end]
    position_ids = reset_ring_attn_position_ids(start, end, packed_seq_lens[0])
    update_ring_attn_params(packed_seq_lens, total_seq_len, ring_attn_group)
    return sequences, attention_mask, position_ids


def reset_position_ids(packed_seq_lens):
    res = []
    for pack in packed_seq_lens:
        position_ids = []
        for length in pack:
            position_ids.extend(list(range(length)))
        res.append(position_ids)
    return torch.tensor(res)


class Runner:
    model_path: str = "/checkpoints/Qwen2.5-7B"
    seed: int = 42
    seq_length: int = 4096

    def run(self):
        max_logits = self.run_flash_attn()
        my_logits = self.run_ring_attn()
        tensor_list = [torch.empty_like(my_logits) for _ in range(self.world_size)]
        dist.all_gather(tensor_list, my_logits)
        my_logits = torch.cat(tensor_list, dim=1)
        if self.rank == 0:
            # The following two has minor differences, WHY?
            print(my_logits)
            print(max_logits)

    @torch.no_grad
    def run_flash_attn(self):
        input_ids, attention_masks, packed_seq_lens = self.data
        position_ids = reset_position_ids(packed_seq_lens)
        input_ids = input_ids.cuda()
        attention_masks = attention_masks.cuda()
        position_ids = position_ids.cuda()
        position_ids.masked_fill_(attention_masks == 0, 1)
        logits = self.model.forward(input_ids, position_ids=position_ids).logits
        return logits.amin(-1)

    @torch.no_grad
    def run_ring_attn(self):
        input_ids, attention_masks, position_ids = convert_ring_attn_params(*self.data, dist.group.WORLD)
        input_ids = input_ids.cuda()
        attention_masks = attention_masks.cuda()
        position_ids = position_ids.cuda()
        position_ids.masked_fill_(attention_masks == 0, 1)
        if self.rank == 7:
            import pdb; pdb.set_trace()
        dist.barrier()
        logits = self.model.forward(input_ids, position_ids=position_ids).logits
        return logits.amin(-1)

    @cached_property
    def data(self):
        input_ids = [[1, 4, 7, 2, 9, 3, 8, 5, 0, 6, 
                     2, 5, 8, 1, 4, 7, 3, 9, 0, 6,
                     5, 8, 2, 4, 1, 7, 3, 9, 0, 6,
                     2, 5]]
        attention_masks = [[1] * 5 + [2] * 8 + [3] * 7 + [4] * 10 + [5] * 2]
        packed_seq_lens = [[5, 8, 7, 10, 2]]
        return torch.tensor(input_ids), torch.tensor(attention_masks), torch.tensor(packed_seq_lens)

    @cached_property
    def model(self):
        from transformers import Qwen2ForCausalLM

        model = Qwen2ForCausalLM.from_pretrained(
            self.model_path,
            device_map="cuda",
            torch_dtype=torch.bfloat16,
            attn_implementation="flash_attention_2"
        )
        model.gradient_checkpointing_enable()

        model.eval()
        return model

    @cached_property
    def rank(self):
        return dist.get_rank()
    
    @cached_property
    def world_size(self):
        return dist.get_world_size()


if __name__ == "__main__":
    dist.init_process_group()
    torch.cuda.set_device(dist.get_rank())
    Runner().run()
    dist.destroy_process_group()
```

**Comment by fingertap** [comment]
Is there anything I did wrong?

**Comment by ChaosCodes** [comment]
I have also noticed the similar issue when using OpenRLHF with ring attn, the loss between `ring_attn_size=8` and `ring_attn_size=4` will have a large difference.

**Comment by fingertap** [comment]
Closing this due to inactive response. I ended up using `liger-kernel` to train 32k sequence.
